exonum/src/blockchain/transaction.rs

```diff
@@ -0,0 +1,490 @@
+        if let Some(description) = self.err().and_then(|e| e.description) {
+            res.extend(bool::into_bytes(true));
+            res.extend(String::into_bytes(description));
+        } else {
+            res.extend(bool::into_bytes(false));
+        }
+        res
+    }
+
+    fn from_bytes(bytes: Cow<[u8]>) -> Self {
+        let main_part = u16::from_bytes(Cow::Borrowed(&bytes));
+        let description = if bool::from_bytes(Cow::Borrowed(&bytes[2..3])) {
+            Some(String::from_bytes(Cow::Borrowed(&bytes[3..])))
+        } else {
+            None
+        };
+
+        match main_part {
+            value @ 0...MAX_ERROR_CODE => Err(TransactionError::code(value as u8, description)),
+            TRANSACTION_STATUS_OK => Ok(()),
+            TRANSACTION_STATUS_PANIC => Err(TransactionError::panic(description)),
+            value => panic!("Invalid TransactionResult value: {}", value),
+        }
+    }
+}
+
+fn status_as_u16(status: &TransactionResult) -> u16 {
+    match *status {
+        Ok(()) => TRANSACTION_STATUS_OK,
+        Err(ref e) => {
+            match e.error_type {
+                TransactionErrorType::Panic => TRANSACTION_STATUS_PANIC,
+                TransactionErrorType::Code(c) => u16::from(c),
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use futures::sync::mpsc;
+
+    use std::collections::BTreeMap;
+    use std::sync::Mutex;
+
+    use super::*;
+    use crypto;
+    use blockchain::Blockchain;
+    use storage::{Database, MemoryDB, Entry};
+    use node::ApiSender;
+    use helpers::{ValidatorId, Height};
+
+    lazy_static! {
+        static ref EXECUTION_STATUS: Mutex<ExecutionResult> = Mutex::new(Ok(()));
+    }
+
+    #[test]
+    fn errors_conversion() {
+        let execution_errors = [
+            ExecutionError::new(0),
+            ExecutionError::new(255),
+            ExecutionError::with_description(1, "".to_owned()),
+            ExecutionError::with_description(1, "Terrible failure".to_owned()),
+        ];
+
+        for execution_error in &execution_errors {
+            let transaction_error: TransactionError = execution_error.clone().into();
+            assert_eq!(execution_error.description, transaction_error.description);
+
+            let code = match transaction_error.error_type {
+                TransactionErrorType::Code(c) => c,
+                _ => panic!("Unexpected transaction error type"),
+            };
+            assert_eq!(execution_error.code, code);
+        }
+    }
+
+    #[test]
+    fn transaction_results_round_trip() {
+        let results = [
+            Ok(()),
+            Err(TransactionError::panic(None)),
+            Err(TransactionError::panic(Some("".to_owned()))),
+            Err(TransactionError::panic(
+                Some("Panic error description".to_owned()),
+            )),
+            Err(TransactionError::code(0, None)),
+            Err(TransactionError::code(
```

Comment:
Nitpick: Would it make sense to introduce an auxiliary function returning `Err(TransactionError::code(x, Some(s.to_owned())))` based on `x` and `s`? All these parentheses make me a bit uncomfortable.